packages/std/src/addresses.rs

```diff
@@ -60,6 +60,12 @@ impl Addr {
     pub fn as_bytes(&self) -> &[u8] {
         self.0.as_bytes()
     }
+
+    /// Utility to downgrading `Addr` to underlying data.
+    #[inline]
+    pub fn into_inner(self) -> String {
+        self.0
+    }
```

Comment:
This is the same as `String::from(addr)`. Do we need this API?

Reply:
Ok, I basically confused myself looking to long for this code and somehow missed it while I needed it. The slight problem is, that this is undocummented function - I would consider changing implementation so instead of implementing `From<addr> for String` there is `Into<String> for Addr` - looks like silly change, but implementing traits on own types generates way better documentation that implementing mirrored traits on external types.

Reply:
We can have an `into_string()` for documentation reasons, that's fine.

Changing From -> Into makes is a step back. From implies Into, but not the other way around. In some situation you need a From implementation (such as the ? operator). It is recommended to implement From and clippy warns you if you don't.

Reply:
I think it's fine and probably idiomatic to have an explicit method for this on the concrete type itself - just for discoverability.

Reply:
We already have `as_str()` for the same reason, which is redundant to the as_ref() implementation.

Reply:
Right, I was somehow sure that those traits are completely mirrored but it obviously doesn't make sense as type elision would get crazy.

Reply:
That doc doesn't mention `into_string` at all, and I'm pretty sure `into_string` is pretty conventional too. Huh.

They talk about wrappers for buffers and the like. Meanwhile, in the std you have [PathBuf](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.into_os_string) and [str](https://doc.rust-lang.org/std/primitive.str.html#method.into_string) that defy the `into_inner` convention.

Reply:
But why would a caller need to know the inner type? Inner is private and can be anything. E.g. a Vector.

Reply:
Yeah, I prefer `into_string` just because it provides information that's actually useful to the caller.

Reply:
This explains my thinking even better than I could:

> When a type wraps a single value to associate it with higher-level semantics, access to the wrapped value should be provided by an into_inner() method.

https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv

We have no wrapping semantics here. Addr and String are different things and there is no semantic connection between the two.

Reply:
@uint `Box<str>` doesn't apply, it performs conversion (`to_owned`) basically, `PathBuf` is a strange exception - `into_inner` seems to be widely used in stdlib https://doc.rust-lang.org/std/index.html?search=into_inner.

@webmaster128 basically I partially agree with this argument. My reasoning about `into_inner` being better than `into_string` (or even `String::from`) is that `into_inner` suggest being free (semantically it promises no conversions are performed). `into_string` obviously might be free, but I would expect that maybe internals aren't actually string, and conversion might be heavy, so maybe I should avoid this. Now I am kind of debating, but I think I agree with you that this type is not purposed to be particular wrapper, and technically it is nothing wrong to change its internal expecting that it wont affect API so I will actually change it this way.

Reply:
That's an excellent argument, but it applies to every `Into` implementation in the ecosystem. Here we have the chance to document how it behaves (if we want to guarantee that behaviour).

Reply:
> @uint `Box<str>` doesn't apply, it performs conversion (`to_owned`) basically, `PathBuf` is a strange exception - `into_inner` seems to be widely used in stdlib https://doc.rust-lang.org/std/index.html?search=into_inner.

Yeah, you're right about `Box<str>`. Most of the types from that search are **generic** wrappers though. The only exception I think are the atomic types.